datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedTaskCountEstimator.java

```diff
@@ -27,6 +41,22 @@
    */
   public int getTaskCount(ClusterThroughputInfo throughputInfo, List<String> assignedPartitions,
       List<String> unassignedPartitions) {
-    throw new NotImplementedException();
+    Validate.notNull(throughputInfo, "null throughputInfo");
+    Validate.notNull(assignedPartitions, "null assignedPartitions");
+    Validate.notNull(unassignedPartitions, "null unassignedPartitions");
+
+    Map<String, PartitionThroughputInfo> throughputMap = throughputInfo.getPartitionInfoMap();
+
+    Set<String> allPartitions = new HashSet<>(assignedPartitions);
+    allPartitions.addAll(unassignedPartitions);
+
+    // total throughput in KB/sec
+    int totalThroughput = throughputMap.entrySet().stream().
+        filter(e -> allPartitions.contains(e.getKey())).mapToInt(e -> e.getValue().getBytesInKBRate()).sum();
+    LOG.debug("Total throughput in all partitions: {}KB/sec", totalThroughput);
+
+    int taskCountEstimate = (int) Math.ceil((double) totalThroughput / (TASK_CAPACITY_MBPS_DEFAULT * 1024));
```

Comment:
Along with `ceil`, I feel there is a need for some additional buffer. This task estimator should keep some scope for growth rather than tightly packing all the partitions in the task.

Since the kafka limit is 4MB, I think one way to get additional tasks is by either considering the default to be 3MB, which in other words means 75% throughput utilization and additional 25% buffer to grow, without increasing the task itself. 

The distribution will have some gaps/holes and it demands for some approximation.

The task estimator will need some smartness in these aspects

Reply:
The way I see it, the buffer for growth will be reserved by elastic task mechanism. We already have that there, right? Do we need to bring it here as well?